lib/bson/parser/deserializer.js

```diff
@@ -23,8 +23,11 @@ var deserialize = function(buffer, options, isArray) {
     (buffer[index + 2] << 16) |
     (buffer[index + 3] << 24);
 
+  const sizeCheck = options.allowObjectSmallerThanBufferSize
+    ? buffer.length < size
+    : buffer.length !== size;
   // Ensure buffer is valid size
-  if (size < 5 || buffer.length !== size || size + index > buffer.length) {
+  if (size < 5 || sizeCheck || size + index > buffer.length) {
```

Comment:
I feel like we can do better here and emit a different `TypeError` for each condition, this would be far more helpful for our users?